website/docs/configuration/config-reference.md
---
sidebar_position: 10
---

# Configuration Reference

The following reference applies to the configuration files of the CEEMS API server,
CEEMS LB, and web configuration. CEEMS uses Prometheus'
[client config](https://pkg.go.dev/github.com/prometheus/common/config#HTTPClientConfig)
to configure HTTP clients. Thus, most of the configuration used to configure HTTP
clients resembles that of Prometheus'. The configuration reference has also been
inspired by Prometheus documentation.

The file is written in [YAML format](https://en.wikipedia.org/wiki/YAML), defined
by the scheme described below. Brackets indicate that a parameter is optional.
For non-list parameters, the value is set to the specified default.

:::tip[TIP]

Some of the configuration files used by the CEEMS components support expanding
environment variables referenced inside the YAML file. This is not enabled by default
and it can be enabled using CLI flag `--*.expand-env-vars`. For example, the environment
variables in the configuration file to CEEMS API server can be expanded by using CLI flag,
`--config.file.expand-env-vars`. Note that when expanding environment variables is enabled,
`$` can be escaped using `$$`.

It is important to note that expanding environment variables is
**not supported for web configuration file.**

:::

Generic placeholders are defined as follows:

* `<boolean>`: a boolean that can take the values `true` or `false`
* `<duration>`: a duration matching the regular expression `((([0-9]+)y)?(([0-9]+)w)?(([0-9]+)d)?(([0-9]+)h)?(([0-9]+)m)?(([0-9]+)s)?(([0-9]+)ms)?|0)`, e.g., `1d`, `1h30m`, `5m`, `10s`
* `<date>`: a date in the format `YYYY-MM-DD`
* `<filename>`: a valid path in the current working directory
* `<float>`: a floating-point number
* `<host>`: a valid string consisting of a hostname or IP followed by an optional port number
* `<int>`: an integer value
* `<path>`: a valid URL path
* `<scheme>`: a string that can take the values `http` or `https`
* `<secret>`: a regular string that is a secret, such as a password
* `<string>`: a regular string
* `<size>`: a size in bytes, e.g., `512MB`. A unit is required. Supported units: B, KB, MB, GB, TB, PB, EB.
* `<idname>`: a string matching the regular expression `[a-zA-Z_-][a-zA-Z0-9_-]*`. Any other unsupported character in the source label should be converted to an underscore
* `<managername>`: a string that identifies the resource manager. Currently accepted values are `slurm`, `openstack` and `k8s`.
* `<updatername>`: a string that identifies the updater type. Currently accepted values are `tsdb`.
* `<promql_query>`: a valid PromQL query string.
* `<lbstrategy>`: a valid load balancing strategy. Currently accepted values are `round-robin` and `least-connection`.
* `<object>`: a generic object

The other placeholders are specified separately.

## `<ceems_api_server>`

The following shows the reference for the CEEMS API server configuration. A valid sample configuration file can be found in the [repository](https://github.com/@ceemsOrg@/@ceemsRepo@/blob/main/build/config/ceems_api_server/ceems_api_server.yml).

```yaml
# Configuration file to configure CEEMS API server
#
# This config file has following sections:
#  - `ceems_api_server`: Core configuration of CEEMS API server
#  - `clusters`: Configuration of clusters that are being monitored
#  - `updaters`: Configuration of updaters
#
---
# CEEMS API Server and data config
ceems_api_server:
  # Data-related configuration of the CEEMS API server. This configuration concerns
  # the locations where data will be saved, frequency of data updates, etc.
  #
  data:
    [ <data_config> ]

  # HTTP web admin-related config for CEEMS API server
  #
  admin:
    [ <admin_config> ]

# A list of clusters from which the CEEMS API server will fetch the compute units.
# 
# Each cluster must provide a unique `id`. The `id` will enable CEEMS to identify 
# different clusters in a multi-cluster setup. This `id` must be consistent throughout 
# all CEEMS components.
# 
clusters:
  [ - <cluster_config> ... ]
  
    
# A list of updaters that will be used to update the compute unit metrics. This update 
# step can be used to update the aggregate metrics of each compute unit in real time
# or to add complementary information to the compute units from on-premise third-party 
# services.
#
# Currently, only the TSDB updater is supported. The compute unit aggregate metrics can be
# updated from TSDB (Prometheus/VM) instances.
#
updaters:
  [ - <updater_config> ... ]
  
```

### `<data_config>`

A `data_config` allows configuring the database settings of the CEEMS API server.

```yaml
# Path at which the CEEMS API server data will be stored.
# If a relative path is used, it will be resolved based on the current working directory.
#
[ path: <filename> | default = data ]

# The duration to retain the data in the database. Units older than this duration will be
# purged from the database. 
#
# In the case of global usage stats, if the last activity on a given project/user 
# combination is older than this period, those stats will be purged from the database.
#
# Units Supported: y, w, d, h, m, s, ms.
#
[ retention_period: <duration> | default = 30d ]

# Units data will be fetched at this interval. CEEMS will pull the units from the 
# underlying resource manager at this frequency into its own database.
#
# Units Supported: y, w, d, h, m, s, ms.
#
[ update_interval: <duration> | default = 15m ]

# Units data will be fetched from this date. If left empty, units will be fetched
# from the current day's midnight.
#
# Format Supported: 2025-01-01.
#
[ update_from: <date> | default = today ]

# Units data will be fetched at this interval when fetching historical data. For
# example, if `update_from` is set to a date in the past, units will be fetched
# for every `max_update_interval` period until we reach the current time, and then
# they will be fetched every `update_interval` time.
#
# Units Supported: y, w, d, h, m, s, ms.
#
[ max_update_interval: <duration> | default = 1h ]

# Time zone to be used when storing times of different events in the DB.
# It takes a value defined in IANA (https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)
# like `Europe/Paris`
# 
# A special value `Local` can be used to use the server's local time zone.
#
[ time_zone: <string> | default = Local ]

# The CEEMS API server is capable of creating database backups using the SQLite backup API. Created
# database backups will be saved to this path. NOTE that for huge databases, this backup can take 
# a considerable amount of time. 
#
# Use a different disk device than `ceems_api_server.data.path` to achieve 
# fault tolerance.
#
# If the path is empty, no backups will be created.
#
[ backup_path: <filename> ]

# The interval at which database backups will be created. 
#
# The minimum allowable interval is `1d`, i.e., 1 day.
#
# Units Supported: y, w, d, h, m, s, ms.
#
[ backup_interval: <duration> | default = 1d ]

```

### `<admin_config>`

The `admin_config` allows configuring the admin users of the CEEMS API server.

```yaml
# List of users that will have admin privileges for accessing the CEEMS API server
#
# These users will have full access to the database and can query stats of any user/project.
#
# In addition, it is possible to pull users from Grafana teams and add them to 
# admin users. Check the `grafana` configuration on how to fetch users from Grafana.
#
users:
    [ - <string> ... ]

# Besides setting a static list of admin users using `ceems_api_server.web.admin_users`,
# it is possible to pull the users from a given Grafana instance and update the admin users
# list of the CEEMS API server. This allows operators to add new admins to the CEEMS API server
# without having to restart `ceems_api_server`. 
#
# Typically, one or several Grafana team(s) can be created dedicated to CEEMS admins, and 
# the CEEMS API server will fetch the Grafana team members at the same frequency as compute 
# units.
#
# The web configuration of Grafana can be set in the following section:
#
grafana:
  [ <grafana_config> ]
```

### `<grafana_config>`

A `grafana_config` allows configuring the Grafana client configuration to fetch members of Grafana teams to be added to admin users of the CEEMS API server.

```yaml
# Web URL of the Grafana instance
#
url: <host>

# List of IDs of the Grafana teams from which the members will be synchronized 
# with CEEMS admin users
#
teams_ids:
  - <string> ...

# Sets the `Authorization` header on every API request with the
# configured username and password.
# password and password_file are mutually exclusive.
#
basic_auth:
  [ username: <string> ]
  [ password: <secret> ]
  [ password_file: <string> ]

# Sets the `Authorization` header on every API request with
# the configured credentials.
#
authorization:
  # Sets the authentication type of the request.
  [ type: <string> | default = Bearer ]
  # Sets the credentials of the request. It is mutually exclusive with
  # `credentials_file`.
  [ credentials: <secret> ]
  # Sets the credentials of the request with the credentials read from the
  # configured file. It is mutually exclusive with `credentials`.
  [ credentials_file: <filename> ]

# Optional OAuth 2.0 configuration.
# Cannot be used at the same time as basic_auth or authorization.
#
oauth2: 
  [ <oauth2> ]

# Configure whether scrape requests follow HTTP 3xx redirects.
[ follow_redirects: <boolean> | default = true ]

# Whether to enable HTTP2.
[ enable_http2: <boolean> | default = true ]

# Configures the API request's TLS settings.
#
tls_config:
  [ <tls_config> ]

# List of headers that will be passed in the API requests to the server.
# Authentication-related headers may be configured in this section. The header name
# must be configured as a key, and the header value supports three different types of 
# headers: values, secrets, and files.
#
# The difference between values and secrets is that secrets will be redacted
# in server logs, whereas values will be emitted in the logs.
#

CHANGELOG.md
# Changelog

## 0.16.0 / 2026-09-09

### Breaking Changes

#### cacct

The configuration file of `cacct` has some breaking changes. Users need to migrate the
section `tsdb.queries` in `cacct`'s configuration file. A key-value configuration has been
changed to list of objects to have more flexibility and to support `instant` type queries
in the future. For instance, if the old configuration file is as follows:

```yaml
tsdb:
  web:
    url: http://localhost:9090
  queries:
    cpu_usage: uuid:ceems_cpu_usage:ratio_irate{uuid=~"%s"}
    cpu_mem_usage: uuid:ceems_cpu_memory_usage:ratio{uuid=~"%s"}
```

It must be modified as follows:

```yaml
tsdb:
  web:
    url: http://localhost:9090
  queries:
    - name: cpu_usage
      title: "CPU Usage"
      help: "CPU usage during the duration of the job"
      query: uuid:ceems_cpu_usage:ratio_irate{uuid=~`{{.UUIDs}}`}
      kind: range
    - name: cpu_mem_usage
      title: "CPU Memory Usage"
      help: "CPU memory usage during the duration of the job"
      query: uuid:ceems_cpu_memory_usage:ratio{uuid=~`{{.UUIDs}}`}
      kind: range
```

where `name`, `query` and `kind` are mandatory keys where as `help` and `title` are
optional. Currently, only `range` type queries are supported for `kind` key. More details
on the full configuration file can be found in [docs](https://ceems-dev.github.io/ceems/docs/configuration/cacct).

### List of PRs

- [FEAT] feat: Support filtering results returned by API server [#596](https://github.com/ceems-dev/ceems/pull/596) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] Ensure ceems_tool accounts multiple instances of emission collectors [#593](https://github.com/ceems-dev/ceems/pull/593) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [BREAKING] feat: Refactor cacct app to configure TSDB custom queries [#585](https://github.com/ceems-dev/ceems/pull/585) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] Bump dependencies [#583](https://github.com/ceems-dev/ceems/pull/583), [#584](https://github.com/ceems-dev/ceems/pull/584), [#586](https://github.com/ceems-dev/ceems/pull/586), [#588](https://github.com/ceems-dev/ceems/pull/588), [#589](https://github.com/ceems-dev/ceems/pull/589), [#590](https://github.com/ceems-dev/ceems/pull/590), [#591](https://github.com/ceems-dev/ceems/pull/591), [#594](https://github.com/ceems-dev/ceems/pull/594), [#595](https://github.com/ceems-dev/ceems/pull/595) ([@dependabot](https://github.com/dependabot))

## 0.15.1 / 2026-08-15

- [MAINT] refactor: Support user supplied SM count for NVIDIA GPUs [#582](https://github.com/ceems-dev/ceems/pull/582) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] Bump dependencies [#581](https://github.com/ceems-dev/ceems/pull/581) ([@dependabot](https://github.com/dependabot))

## 0.15.0 / 2026-08-04

- [MAINT] Document the constraints around TSDB scrape interval and CEEMS API server's update interval [#575](https://github.com/ceems-dev/ceems/pull/575) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [FEAT] feat: Support querying all users in cacct for admins [#556](https://github.com/ceems-dev/ceems/pull/556) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] Add trusted IP prefixes to API server CLI args [#554](https://github.com/ceems-dev/ceems/pull/554) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] Bump dependencies [#552](https://github.com/ceems-dev/ceems/pull/552), [#560](https://github.com/ceems-dev/ceems/pull/560), [#561](https://github.com/ceems-dev/ceems/pull/561), [#562](https://github.com/ceems-dev/ceems/pull/562), [#563](https://github.com/ceems-dev/ceems/pull/563), [#564](https://github.com/ceems-dev/ceems/pull/564), [#565](https://github.com/ceems-dev/ceems/pull/565), [#567](https://github.com/ceems-dev/ceems/pull/567), [#568](https://github.com/ceems-dev/ceems/pull/568), [#569](https://github.com/ceems-dev/ceems/pull/569), [#570](https://github.com/ceems-dev/ceems/pull/570), [#571](https://github.com/ceems-dev/ceems/pull/571), [#572](https://github.com/ceems-dev/ceems/pull/572), [#576](https://github.com/ceems-dev/ceems/pull/576), [#577](https://github.com/ceems-dev/ceems/pull/577), [#578](https://github.com/ceems-dev/ceems/pull/578), [#579](https://github.com/ceems-dev/ceems/pull/579), [#580](https://github.com/ceems-dev/ceems/pull/580) ([@dependabot](https://github.com/dependabot))

## 0.14.4 / 2026-07-05

- [CI] Fix app token creation in reusable workflow [#550](https://github.com/ceems-dev/ceems/pull/550) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] refactor: Misc improvements and fixes for API server [#549](https://github.com/ceems-dev/ceems/pull/549) ([@mahendrapaipuri](https://github.com/mahendrapaipuri))
- [MAINT] Bump dependencies [#545](https://github.com/ceems-dev/ceems/pull/545), [#546](https://github.com/ceems-dev/ceems/pull/546), [#547](https://github.com/ceems-dev/ceems/pull/547) ([@dependabot](https://github.com/dependabot))


pkg/api/resource/lsf/cli.go
package lsf

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/ceems-dev/ceems/internal/common"
	"github.com/ceems-dev/ceems/internal/osexec"
	"github.com/ceems-dev/ceems/pkg/api/base"
	"github.com/ceems-dev/ceems/pkg/api/models"
)

var (
	// LSF accounting gives memory as 200M, 250.5G and we dont know if it gives without
	// units. However, bjobs return memory as 200 Mbytes.
	// So, regex will capture the number and unit (if exists) and we convert it to bytes.
	// Ref: https://regex101.com/r/ate08j/1
	memRegex = regexp.MustCompile(`([0-9.]+)(?:\s?)([K|M|G|T]?)`)
	toBytes  = map[string]int64{
		"K": 1024,
		"M": 1024 * 1024,
		"G": 1024 * 1024 * 1024,
		"T": 1024 * 1024 * 1024 * 1024,
		"Z": 1024 * 1024 * 1024 * 1024 * 1024,
	}
)

// Run preflights for CLI execution mode.
func preflightsCLI(lsf *lsfScheduler) error {
	// We hit this only when fetch mode is bacct and bjobs commands
	// Assume execMode is always native
	lsf.fetchMode = cliMode
	lsf.logger.Debug("Using LSF CLI commands")

	// If no bacct path is provided, assume it is available on PATH
	if lsf.cluster.CLI.Path == "" {
		path, err := exec.LookPath("bacct")
		if err != nil {
			lsf.logger.Error("Failed to find LSF utility executables on PATH", "err", err)

			return err
		}

		lsf.cluster.CLI.Path = filepath.Dir(path)
	} else {
		// Check if lsf binary directory exists at the given path
		_, err := os.Stat(lsf.cluster.CLI.Path)
		if err != nil {
			lsf.logger.Error("Failed to open LSF bin dir", "path", lsf.cluster.CLI.Path, "err", err)

			return err
		}
	}

	return nil
}

// Parse bacct command output and return batchjob slice.
func parseBacctCmdOutput(bacctOutput string, start time.Time, end time.Time, jobsCache *sync.Map) []models.Unit {
	// Get current location
	loc := end.Location()

	// Update period
	intStartTS := start.UnixMilli()
	intEndTS := end.UnixMilli()

	// Start with a default delimiter
	delimiter := "------------------------------------------------------------------------------"
	// Find the line separator
	sepMatches := regexSeparator.FindAllStringSubmatch(bacctOutput, 1)
	if len(sepMatches) > 0 && len(sepMatches[0]) > 0 {
		delimiter = sepMatches[0][0]
	}

	// Split output to get per job record
	bacctOutputRecords := strings.Split(bacctOutput, delimiter)

	jobs := make([]models.Unit, len(bacctOutputRecords))

	wg := &sync.WaitGroup{}
	wg.Add(len(bacctOutputRecords))

	for irecord, record := range bacctOutputRecords {
		go func(i int, r string) {
			var jobStat models.Unit

			matches := regexJobDetails.FindStringSubmatch(r)

			// Ignore if no matches found
			if len(matches) == 0 {
				wg.Done()

				return
			}

			// Get all named captured groups
			components := make(map[string]string)
			for i, name := range regexJobDetails.SubexpNames() {
				components[name] = matches[i]
			}

			// Ignore jobs that never ran
			if components["start_time"] == "" {
				wg.Done()

				return
			}

			// Get user's job group, ncpus, ngpus from cache map that we fetched from bjobs
			// If we managed to parse bacct output, the nnodes, ncpus, ngpus, mem and nodelist
			// will be overwritten by bacct output. This is desirable as jobs can be
			// modified between last scrape and its termination. So, we get last updated
			// job info
			components["group"] = ""

			var (
				nnodes, ncpus, ngpus int
				nodelist             string
				mem                  int64
			)

			if jobsCache != nil {
				if val, ok := jobsCache.LoadAndDelete(components["jobid"]); ok {
					if u, ok := val.(jobAttributes); ok {
						components["group"] = u.userGroup
						nnodes = u.numNodes
						ncpus = u.numCPUs
						ngpus = u.numGPUs
						mem = u.mem
						nodelist = u.nodelist
					}
				}
			}

			// Convert time strings to configured time location
			eventTS := make(map[string]int64, 3)

			for _, c := range []string{"submit_time", "start_time", "end_time"} {
				// bacct does not include tz in the time format. So, always assume the
				// location to be the current one that API server is running on.
				t, err := time.ParseInLocation(bacctTimeLayoout, components[c], loc)
				if err == nil {
					components[c] = t.Format(base.DatetimezoneLayout)
				}

				eventTS[c] = common.TimeToTimestamp(base.DatetimezoneLayout, components[c])
			}

			// Get CPU allocations
			allocCPUs := strings.Fields(strings.ReplaceAll(strings.ReplaceAll(components["nodelist"], "<", ""), ">", ""))
			if len(allocCPUs) > 0 {
				ncpus = len(allocCPUs)

				// Remove duplicates to get allocated nodes
				slices.Sort(allocCPUs)
				allocNodes := slices.Compact(allocCPUs)
				nnodes = len(allocNodes)
				nodelist = strings.Join(allocNodes, "|")
			}

			// Get memory allocations
			var (
				memString string
				memEff    float64
			)

			for l := range strings.SplitSeq(components["accounting"], "\n") {
				// Split line into columns
				fields := strings.Fields(strings.TrimSpace(l))

				// If there are 7 columns that 5th one is the memory usage
				if len(fields) == 7 && !strings.Contains(l, "MEM") {
					memString = fields[5]
				} else if len(fields) == 2 && !strings.Contains(l, "MEM_EFFICIENCY") {
					memEff, _ = strconv.ParseFloat(strings.TrimSuffix(fields[1], "%"), 64)
				}
			}

			// If mem is not empty string, convert the units [K|M|G|T] into numeric bytes
			// The following logic covers the cases when memory is of form 200M, 250.5G
			// and also without unit eg 20000, 40000. When there is no unit we assume
			// it is already in bytes
			memStringMatches := memRegex.FindStringSubmatch(memString)

			if len(memStringMatches) >= 2 {
				memFloat, err := strconv.ParseFloat(memStringMatches[1], 64)
				if err == nil {
					if len(memStringMatches) == 3 {
						if unitConv, ok := toBytes[memStringMatches[2]]; ok {
							mem = int64(memFloat) * unitConv

							// If memEff is not zero, convert real memory usage
							// into memory allocation
							if memEff > 0 {
								mem = int64(100.0 * float64(mem) / memEff)
							}
						}
					}
				}
			}

			var (
				allocGPUs          []string
				hostname, gpuIndex string
			)

			for l := range strings.SplitSeq(components["gpu_alloc"], "\n") {
				// Ignore empty and header lines
				if l == "" || strings.Contains(l, "HOST") {
					continue
				}

				// Split line into columns
				fields := strings.Fields(strings.TrimSpace(l))

				// If there are 11 columns, there is a hostname in the column and GPU
				// ordinal is at 3rd position.
				// If there is no hostname, we need to use previously found hostname and
				// GPU ordinal will be at 2nd position.
				if len(fields) == 11 {
					hostname = fields[0]
					gpuIndex = fields[2]
				} else if len(fields) == 10 {
					gpuIndex = fields[1]
				}

				allocGPUs = append(allocGPUs, fmt.Sprintf("%s:%s", hostname, gpuIndex))
			}

			// Remove duplicates
			if len(allocGPUs) > 0 {
				slices.Sort(allocGPUs)
				allocGPUs = slices.Compact(allocGPUs)
				ngpus = len(allocGPUs)
			}

			// Assume job's elapsed time during this interval overlaps with interval's
			// boundaries
			startMark := intStartTS
			endMark := intEndTS

			// If job has already finished in the past we need to get boundaries from
			// job's start and end time. This case should not arrive in production as
			// there is no reason LSF gives us the jobs that have finished in the past
			// that do not overlap with interval boundaries
			if eventTS["end_time"] > 0 && eventTS["end_time"] < intStartTS {
				startMark = eventTS["start_time"]
				endMark = eventTS["end_time"]

				goto elapsed
			}

			// If job has started **after** start of interval, we should mark job's start
			// time as start of elapsed time
			if eventTS["start_time"] > intStartTS {
				startMark = eventTS["start_time"]
			}

			// If job has ended before end of interval, we should mark job's end time
			// as elapsed end time.
			if eventTS["end_time"] > 0 && eventTS["end_time"] < intEndTS {
				endMark = eventTS["end_time"]
			}

		elapsed:
			// Get elapsed time of job in this interval in seconds
			elapsedSeconds := (endMark - startMark) / 1000

			// Get cpuSeconds and gpuSeconds of the current interval
			var cpuSeconds, gpuSeconds int64

			cpuSeconds = int64(ncpus) * elapsedSeconds
			gpuSeconds = int64(ngpus) * elapsedSeconds

			// Get cpuMemSeconds and gpuMemSeconds of current interval in MB
			var cpuMemSeconds, gpuMemSeconds int64
			if mem > 0 {
				cpuMemSeconds = mem * elapsedSeconds / toBytes["M"]
			} else {
				cpuMemSeconds = elapsedSeconds
			}

			// Currently we use walltime as GPU mem time. This wont be a correct proxy
			// if MIG is enabled in GPUs where different portions of memory can be
			// allocated
			if ngpus > 0 {
				gpuMemSeconds = elapsedSeconds
			}

			// Allocation
			allocation := models.Allocation{
				"nodes": nnodes,
				"cpus":  ncpus,
				"mem":   mem,
				"gpus":  ngpus,
			}

			// Tags
			tags := models.Tag{
				"queue":       components["queue"],
				"res_req":     components["res_req"],
				"exit_status": components["exit_status"],
				"nodelist":    nodelist,
				"workdir":     components["cwd"],
			}

			// Make jobStats struct for each job and put it in jobs slice
			jobStat = models.Unit{
				ResourceManager: lsfBatchScheduler,
				UUID:            components["jobid"],
				Project:         components["project"],
				User:            components["user"],
				Group:           components["group"],
				CreatedAt:       components["submit_time"],
				StartedAt:       components["start_time"],
				EndedAt:         components["end_time"],
				CreatedAtTS:     eventTS["submit_time"],
				StartedAtTS:     eventTS["start_time"],
				EndedAtTS:       eventTS["end_time"],
				Elapsed:         common.Timespan(time.Duration(eventTS["end_time"]-eventTS["start_time"]) * time.Second / 1e3).Format("15:04:05"),
				State:           components["status"],
				Allocation:      allocation,
				TotalTime: models.MetricMap{
					"walltime":         models.JSONFloat(elapsedSeconds),
					"alloc_cputime":    models.JSONFloat(cpuSeconds),
					"alloc_cpumemtime": models.JSONFloat(cpuMemSeconds),
					"alloc_gputime":    models.JSONFloat(gpuSeconds),
					"alloc_gpumemtime": models.JSONFloat(gpuMemSeconds),
				},
				Tags: tags,
			}

			jobLock.Lock()

			jobs[i] = jobStat

			jobLock.Unlock()
			wg.Done()
		}(irecord, record)
	}

	wg.Wait()

	// Remove empty jobs
	var validJobs []models.Unit

	for _, job := range jobs {
		if job.UUID != "" {
			validJobs = append(validJobs, job)
		}
	}

	return validJobs
}

// runBacctCmd executes bacct command and return output.
func (s *lsfScheduler) runBacctCmd(ctx context.Context, start, end time.Time) ([]byte, error) {
	// bacct path
	bacctPath := filepath.Join(s.cluster.CLI.Path, "bacct")

	// Setup any provided env vars
	var env []string
	for name, value := range s.cluster.CLI.EnvVars {
		env = append(env, fmt.Sprintf("%s=%s", name, value))
	}

	// Get jobs of ALL users in machine readable format
	args := []string{
		"-u", "all", "-l", "-gpu", "-UF",
		"-C", fmt.Sprintf("%s,%s", start.Format(lsfCommandsTimeLayout), end.Format(lsfCommandsTimeLayout)),
	}

	return osexec.ExecuteContext(ctx, bacctPath, args, env)
}

// Parse bjobs command output and return batchjob slice.
func parseBjobsCmdOutput(bjobsOutput []byte, start time.Time, end time.Time, jobsCache *sync.Map) ([]models.Unit, int, error) {
	// Get current location
	loc := end.Location()

	// Update period
	intStartTS := start.UnixMilli()
	intEndTS := end.UnixMilli()

	// Unmarshal command output into LSFJobsList struct
	var bjobs common.LSFJobsList

	err := json.Unmarshal(bjobsOutput, &bjobs)
	if err != nil {
		return nil, 0, fmt.Errorf("failed to unmarshal bjobs output: %w", err)
	}

	jobs := make([]models.Unit, len(bjobs.Records))
	numJobs := 0

	wg := &sync.WaitGroup{}
	wg.Add(len(bjobs.Records))

	for irecord, record := range bjobs.Records {
		go func(i int, r common.LSFJobRecord) {
			var jobStat models.Unit

			// If job state is not running, ignore this job
			// Job states: https://www.ibm.com/docs/en/spectrum-lsf/10.1.0?topic=execution-about-job-states
			if r.Stat != "RUN" {
				wg.Done()

				return
			}

			// Convert time strings to configured time location
			eventTS := make(map[string]int64, 2)

			// Components map to save all the attributes
			components := map[string]string{
				"submit_time": r.SubmitTime,
				"start_time":  r.StartTime,
				"end_time":    r.FinishTime,
			}

			for c, ts := range components {
				// bjobs does not include tz in the time format. So, always assume the
				// location to be the current one that API server is running on.
				t, err := time.ParseInLocation(bjobsTimeLayoout, ts, loc)
				if err == nil {
					components[c] = t.Format(base.DatetimezoneLayout)
				} else {
					components[c] = "N/A"
				}

				eventTS[c] = common.TimeToTimestamp(base.DatetimezoneLayout, components[c])
			}

			// If job has already finished in the past bjobs should not return this job.
			// In case if it does, just ignore it
			if eventTS["end_time"] > 0 && eventTS["end_time"] < intStartTS {
				wg.Done()

				return
			}

			// Get allocated nodes
			allocCPUs := strings.Split(r.AllocSlot, ":")
			slices.Sort(allocCPUs)
			allocNodes := slices.Compact(allocCPUs)
			ncpus := r.NumAllocSlot

			// If mem is not empty string, convert the units [K|M|G|T] into numeric bytes
			// The following logic covers the cases when memory is of form 200 Mbytes, 250.5 Gbtyes
			// and also without unit eg 20000, 40000. When there is no unit we assume
			// it is already in bytes
			var mem int64

			memStringMatches := memRegex.FindStringSubmatch(r.Mem)
			if len(memStringMatches) >= 2 {
				// Get used memory
				memUsed, err := strconv.ParseFloat(memStringMatches[1], 64)
				if err == nil {
					// If units are present, convert to bytes
					if len(memStringMatches) == 3 {
						if unitConv, ok := toBytes[memStringMatches[2]]; ok {
							memUsed = memUsed * float64(unitConv)
						}
					}

					// Get memory efficiency
					memEff, _ := strconv.ParseFloat(strings.TrimSuffix(r.MemEfficiency, "%"), 64)

					// If memory is greater than zero, we can multiply used memory with
					// efficiency to get reserved memory
					if memEff > 0 {
						mem = int64(100 * memUsed / memEff)
					} else {
						mem = int64(memUsed)
					}
				}
			}

			// Get number of GPUs
			ngpus := r.NumGPU

			// Get elapsed time
			// bjobs output returns in format 50 second(s). Need to strip second(s)
			// from string before converting to int
			components["elapsed_time"] = ""

			v, err := strconv.ParseInt(strings.TrimSpace(strings.TrimSuffix(r.Runtime, "second(s)")), 10, 64)
			if err == nil {
				components["elapsed_time"] = common.Timespan(time.Duration(v) * time.Second).Format("15:04:05")
			}

			// Assume job's elapsed time during this interval overlaps with interval's
			// boundaries
			startMark := intStartTS
			endMark := intEndTS

			// If job has started **after** start of interval, we should mark job's start
			// time as start of elapsed time
			if eventTS["start_time"] > intStartTS {
				startMark = eventTS["start_time"]
			}

			// If job has ended before end of interval, we should mark job's end time
			// as elapsed end time.
			if eventTS["end_time"] > 0 && eventTS["end_time"] < intEndTS {
				endMark = eventTS["end_time"]
			}

			// Get elapsed time of job in this interval in seconds
			elapsedSeconds := (endMark - startMark) / 1000

			// Get cpuSeconds and gpuSeconds of the current interval
			var cpuSeconds, gpuSeconds int64

			cpuSeconds = int64(ncpus) * elapsedSeconds
			gpuSeconds = int64(ngpus) * elapsedSeconds

			// Get cpuMemSeconds and gpuMemSeconds of current interval in MB
			var cpuMemSeconds, gpuMemSeconds int64
			if mem > 0 {
				cpuMemSeconds = mem * elapsedSeconds / toBytes["M"]
			} else {
				cpuMemSeconds = elapsedSeconds
			}

			// Currently we use walltime as GPU mem time. This wont be a correct proxy
			// if MIG is enabled in GPUs where different portions of memory can be
			// allocated
			if ngpus > 0 {
				gpuMemSeconds = elapsedSeconds
			}

			// Allocation
			allocation := models.Allocation{
				"nodes": len(allocNodes),
				"cpus":  ncpus,
				"mem":   mem,
				"gpus":  ngpus,
			}

			// Tags
			tags := models.Tag{
				"queue":    r.Queue,
				"res_req":  r.EffectiveResReq,
				"nodelist": strings.Join(allocNodes, "|"),
				"workdir":  r.ExecCWD,
			}

			// Make jobStats struct for each job and put it in jobs slice
			jobStat = models.Unit{
				ResourceManager: lsfBatchScheduler,
				UUID:            r.ID,
				Project:         r.Project,
				User:            r.User,
				Group:           r.UserGroup,
				CreatedAt:       components["submit_time"],
				StartedAt:       components["start_time"],
				EndedAt:         components["end_time"],
				CreatedAtTS:     eventTS["submit_time"],
				StartedAtTS:     eventTS["start_time"],
				EndedAtTS:       eventTS["end_time"],
				Elapsed:         components["elapsed_time"],
				State:           r.Stat,
				Allocation:      allocation,
				TotalTime: models.MetricMap{
					"walltime":         models.JSONFloat(elapsedSeconds),
					"alloc_cputime":    models.JSONFloat(cpuSeconds),
					"alloc_cpumemtime": models.JSONFloat(cpuMemSeconds),
					"alloc_gputime":    models.JSONFloat(gpuSeconds),
					"alloc_gpumemtime": models.JSONFloat(gpuMemSeconds),
				},
				Tags: tags,
			}

			// Push jobStat struct into jobsCache to be used in bacct parsing
			jobAttr := jobAttributes{
				numNodes:  len(allocNodes),
				numCPUs:   ncpus,
				numGPUs:   ngpus,
				userGroup: r.UserGroup,
				nodelist:  strings.Join(allocNodes, "|"),
				mem:       mem,
			}
			if _, exists := jobsCache.LoadOrStore(r.ID, jobAttr); exists {
				jobsCache.Swap(r.ID, jobAttr)
			}

			jobLock.Lock()

			jobs[i] = jobStat
			numJobs += 1

			jobLock.Unlock()
			wg.Done()
		}(irecord, record)
	}

	wg.Wait()

	return jobs, numJobs, nil
}

// runBjobsCmd executes bjobs command and return output.
func (s *lsfScheduler) runBjobsCmd(ctx context.Context) ([]byte, error) {
	// bjobs path
	bjobsPath := filepath.Join(s.cluster.CLI.Path, "bjobs")

	// Setup any provided env vars
	var env []string
	for name, value := range s.cluster.CLI.EnvVars {
		env = append(env, fmt.Sprintf("%s=%s", name, value))
	}

	// Get jobs of ALL users in machine readable format
	args := []string{
		"-u", "all", "-json", "-o", "jobid jobindex user user_group queue job_name proj_name stat alloc_slot nalloc_slot submit_time start_time finish_time run_time effective_resreq gpu_num gpu_alloc mem memlimit mem_efficiency exec_cwd",
	}

	return osexec.ExecuteContext(ctx, bjobsPath, args, env)
}

// Run preflight checks on provided config.
func preflightChecks(s *lsfScheduler) error {
	// // Always prefer REST API mode if configured
	// if clusterConfig.Web.URL != "" {
	// 	return checkRESTAPI(clusterConfig, logger)
	// }
	return preflightsCLI(s)
}
